crates/ruff_python_parser/src/lexer.rs

```diff
@@ -817,25 +792,38 @@ impl<'source> Lexer<'source> {
             }
         };
 
-        Ok(Tok::String {
+        self.value = TokenValue::String {
             value: self.source[TextRange::new(value_start, value_end)]
                 .to_string()
                 .into_boxed_str(),
             kind,
-        })
+        };
+
+        Ok(TokenKind::String)
     }
 
-    // This is the main entry point. Call this function to retrieve the next token.
-    // This function is used by the iterator implementation.
-    pub fn next_token(&mut self) -> LexResult {
+    pub fn next_token(&mut self) -> Token {
+        let next = self.next_token_impl();
+        self.current = match next {
+            Ok(token) => token,
+            Err(error) => {
+                let range = error.location();
+                self.errors.push(error);
+                Token::new(TokenKind::Unknown, range)
+            }
+        };
+        self.current
+    }
```

Comment:
I'm not sure what this will do to performance. It adds some additional branches to every `next_token` call even when the error case is extremely rare. 

I think I would prefer if we instead change all lex methods to simply return `TokenKind` and add a new `error_token(error)` method that returns `TokenKind::Unknown` and pushes the error. I think this has a few advantages

* We move out the error branching from the very hot `next_token` function to the few places where errors can happen and we already have an error branch there (we only pay the cost of an error branch once instead of twice)
* The lexer can now push multiple errors for a single `next_token` call. It even gets the option to push an error, but continue lexing (e.g. push an error for an unterminated string literal but continue lexing)
* We drastically reduce the return type size of all lex functions because I think `LexicalError` is somewhat heavy. It certainly is heavier than a single `TokenKind` :)

Reply:
Yeah, actually I had the same plan but I did this just to simplify the implementation for now. I should've made that clear 😅